### ISSUE TITLE
[ServiceBus] message.dead_letter doesn't set any properties of message which is consistent with the .NET Sdk

### ISSUE THREAD

**Issue by yunhaoling** [Azure/azure-sdk-for-python#9242]
Customer report:

When you are receiving a message and explicitly try to deadletter it (message.dead_letter(description)), the description (string) doesn't show up in the properties of the message even when it is specified.  I tried it with the regular dead_letter and also async dead_letter in Python, both returned these properties of the message:
"properties": {},
When I run the same test using the .NET SDK for Service Bus, it works as expected.  Here are the message properties from .NET
"properties": {
    "DeadLetterReason": "bad"
  }

----------

Repro:

```python
    queue_client = client.get_queue(standard_queue)
    with queue_client.get_receiver(idle_timeout=5, mode=ReceiveSettleMode.PeekLock, prefetch=10) as receiver:

        with queue_client.get_sender() as sender:
            for i in range(10):
                message = Message("Dead lettered message no. {}".format(i))
                sender.send(message)

        count = 0
        messages = receiver.fetch_next()
        while messages:
            for message in messages:
                print_message(message)
                message.dead_letter(description="Testing queue deadletter")
                count += 1
            messages = receiver.fetch_next()

    assert count == 10

    with queue_client.get_deadletter_receiver(idle_timeout=5, mode=ReceiveSettleMode.PeekLock) as receiver:
        count = 0
        for message in receiver:
            print_message(message)
            assert message.user_properties
            assert message.user_properties[b"DeadLetterReason"]
            message.complete()
            count += 1
    assert count == 10
```

**Comment by MichaelWeedon** [comment]
Is there an ETA when this issue might be addressed?

**Comment by KieranBrantnerMagee** [comment]
Hey @MichaelWeedon, my apologies for the slow reply, am digging into this issue and will loop back with ETA once we get a sense for the resolution.

Thanks in advance for all your patience thus far.

**Comment by yunhaoling** [comment]
Hi @MichaelWeedon , sorry for keeping you waiting for such a long time.

I have created a patch wheel for the issue, can you test whether it solves the problem on your side?
[dead-letter-description-patch.zip](https://github.com/Azure/azure-sdk-for-python/files/4315071/dead-letter-description-patch.zip)

(Code change could be found here if you're interested in implementation detail https://github.com/Azure/azure-sdk-for-python/pull/10246)

**Comment by MichaelWeedon** [comment]
@yunhaoling 
installed the patch - getting an exception in dead_letter().
Suspect I am missing something.

  File "/Users/xxxx/projects/venv/lib/python3.6/site-packages/azure/servicebus/aio/async_message.py", line 108, in dead_letter
    raise MessageSettleFailed("reject", e)
azure.servicebus.common.errors.MessageSettleFailed: Failed to reject message. Error: Management request failed: Management request returned status code: 500. Description: b'Unsupported type 0x40 for array. TrackingId:2d0f849e-1528-4642-a8cc-ee6fc813dcd1_G4, SystemTracker:NoSystemTracker, Timestamp:2020-03-11T14:38:23', Data: None

**Comment by yunhaoling** [comment]
@MichaelWeedon,

Which version of uamqp library are you using, can you try installing the the latest uamqp library by `pip install uamqp --upgrade`? -- that should install uamqp v1.2.6 for you.

We're aware of the [issue](https://github.com/Azure/azure-sdk-for-python/issues/7336) `Error: Management request failed: Management request returned status code: 500. Description: b'Unsupported type 0x40 for array.` and made a fix in the uamqp library.

**Comment by MichaelWeedon** [comment]
@yunhaoling 
Already at 1.2.6 of uamqp.

$ pip list | grep uamqp
uamqp                                         1.2.6      
$ pip install uamqp --upgrade
Requirement already up-to-date: uamqp in /Users/xxx/projects/venv/lib/python3.6/site-packages (1.2.6)
Requirement already satisfied, skipping upgrade: certifi>=2017.4.17 in /Users/xxx/projects/venv/lib/python3.6/site-packages (from uamqp) (2019.9.11)
Requirement already satisfied, skipping upgrade: six~=1.0 in /Users/xxx/projects/venv/lib/python3.6/site-packages (from uamqp) (1.11.0)

**Comment by yunhaoling** [comment]
That's interesting, when I test the functionality it works for me.

Which platform are you using, windows/linux/mac?
Are you doing some long-period operation on a single message? It would be great if you could share me with the simplified code to reproduce the `b'Unsupported type 0x40 for array.` issue.

Thanks

**Comment by MichaelWeedon** [comment]
@yunhaoling 
I am on Mac OS.
I pulled the latest sdk samples and they have been reworked and I no longer get the 0x40.
Using example_session_send_receive_with_pool_async.py in async_samples
replace line #32 with the following (sending all message to dead_leter)
                    await message.dead_letter("Bad Message")
 but still no dead letter attributes

**Comment by yunhaoling** [comment]
@MichaelWeedon thanks for the additional context, I would try to reproduce the issue.
Will keep you updated.

**Comment by yunhaoling** [comment]
@MichaelWeedon sorry for not getting back to you sooner.

It took me some time to figure out the difference between session/non-session entity when handling dead letter. I've made another patch, this one works well for both type of entity on my side. Can you help verify the patch solves your problem or not?

[azure_servicebus-0.50.3b2-py2.py3-none-any.zip](https://github.com/Azure/azure-sdk-for-python/files/4357199/azure_servicebus-0.50.3b2-py2.py3-none-any.zip)

**Comment by MichaelWeedon** [comment]
@yunhaoling, installed latest patch and do not see a difference in behavior in the samples or my program - meaning no attributes in dead letter.  Do you have code that illustrates this properly sending the attributes to the dead letter queue?

**Comment by yunhaoling** [comment]
@MichaelWeedon Here is the code I used, please replace `CONN_STR `, `QUEUE_NAME` and `SESSION_ID ` with your own configs.

```python
from azure.servicebus.aio import ServiceBusClient, Message
from azure.servicebus import ReceiveSettleMode
import asyncio

import logging
logging.basicConfig(level=logging.INFO)

_logger = logging.getLogger(__name__)


CONN_STR = "<YOUR CONNECTION STR>"
QUEUE_NAME = "<YOUR QUEUE NAME>"
SESSION_ID = "<YOUR SESSION ID>"


def print_message(message):
    _logger.info("Receiving: {}".format(message))
    _logger.debug("Time to live: {}".format(message.time_to_live))
    _logger.debug("Sequence number: {}".format(message.sequence_number))
    _logger.debug("Enqueue Sequence number: {}".format(message.enqueue_sequence_number))
    _logger.debug("Partition ID: {}".format(message.partition_id))
    _logger.debug("Partition Key: {}".format(message.partition_key))
    _logger.info("User Properties: {}".format(message.user_properties))
    _logger.debug("Annotations: {}".format(message.annotations))
    _logger.debug("Delivery count: {}".format(message.header.delivery_count))
    try:
        _logger.debug("Locked until: {}".format(message.locked_until))
        _logger.debug("Lock Token: {}".format(message.lock_token))
    except TypeError:
        pass
    _logger.debug("Enqueued time: {}".format(message.enqueued_time))


async def main():
    client = ServiceBusClient.from_connection_string(conn_str=CONN_STR)

    queue_client = client.get_queue(QUEUE_NAME)
    async with queue_client.get_receiver(session=SESSION_ID, idle_timeout=10, mode=ReceiveSettleMode.PeekLock, prefetch=10) as receiver:
        async with queue_client.get_sender() as sender:
            for i in range(10):
                message = Message("Dead lettered message no. {}".format(i))
                message.session_id = SESSION_ID
                await sender.send(message)

        count = 0
        messages = await receiver.fetch_next()
        while messages:
            for message in messages:
                print_message(message)
                await message.dead_letter(description="Testing queue deadletter")
                count += 1
            messages = await receiver.fetch_next()

    assert count == 10

    async with queue_client.get_deadletter_receiver(idle_timeout=10, mode=ReceiveSettleMode.PeekLock) as receiver:
        count = 0
        async for message in receiver:
            print_message(message)
            assert message.user_properties[b'DeadLetterReason'] == b'Testing queue deadletter'
            assert message.user_properties[b'DeadLetterErrorDescription'] == b'Testing queue deadletter'
            await message.complete()
            count += 1
    assert count == 10

loop = asyncio.get_event_loop()
loop.run_until_complete(main())
```

**Comment by MichaelWeedon** [comment]
@yunhaoling,  Thanks.  Able to get it to work.  One caveat is I cannot get the reason (i.e description) to be set if I process the dead_letter within a handle exception.  Meaning if while processing the message an exception occurs, I want to dead_letter the message using the exception as the reason. In this scenario, no reason is set - thoughts?

**Comment by yunhaoling** [comment]
Hi @MichaelWeedon , can you share me with code that reproduce the issue? -- dead_letter message when an exceptions occurs and use the exception as the reason.

**Comment by MichaelWeedon** [comment]
@yunhaoling Upon further investigation, it appears the problem is corrected for all cases. For some reason my pycharm environment was not picking up the newer sdk you supplied.
That said, when can this me made generally available?

**Comment by yunhaoling** [comment]
Hi @MichaelWeedon , thanks for the feedback, glad it works for you.

Currently we're triaging some other ServiceBus issues and try to get them resolved.
I can't guarantee the time for the next release but it's definitely in our plan.

We'll keep you updated about the release plan, thanks for your patience in advance.

**Comment by yunhaoling** [comment]
Hi @MichaelWeedon , apologize for the late response. We've released azure-servicebus v0.50.3 today which should have the issue addressed. Please update the library.

One thing I want to note is that v0.50.3 would be our final release for v0.50.

The good news is you can try using our latest preview library, we're putting all our efforts into [azure-servicebus v7.0.0](https://pypi.org/project/azure-servicebus/7.0.0b2/) (which is currently in b2) to create a more user friendly and productive library. As we have limited cycles, we decided not to maintain the old preview (v0.50) library which will never go GA.

We've provided [migration guide on v0.50 to v7](https://github.com/Azure/azure-sdk-for-python/blob/master/sdk/servicebus/azure-servicebus/migration_guide.md) to help get you onboard.

Apologize for the inconvenience we bring to you and thanks for your understanding (and for sure your collaboration on fixing this issue!).

We're looking forward to you trying our latest preview library and feedback is most welcome.

**Comment by MichaelWeedon** [comment]
@yunhaoling, Thanks for the update.  I tried the 7.0.0b2 using the newer api/examples and the same problem occurs - Users Errors in the portal.  I see there is a comment in the sample code disclaiming there maybe user errors in the portal.  I've attempted to find an async method for processing service bus messages without such errors, but have found no way to re-engineer the code to eliminate this occurrence in either the 0.5.2 or the 7.0.0b2 apis.
